processor.go

```diff
@@ -442,6 +442,123 @@ func callInterfaceFunc(handler, input interface{}, streaming bool) (intf interfa
 	return
 }
 
+// ProcessClientStreamRequest allows clients to push multiple messages
+// asynchronously to the same service with the same connection. Unlike in
+// ProcessClientRequest, we take a channel of inputs that can be filled and will
+// subsequently call the service with any new messages received in the channel.
+func (p *ServiceProcessor) ProcessClientStreamRequest(req *http.Request, path string,
+	clientInputs chan []byte) ([]byte, *StreamingTunnel, error) {
+
+	outChan := make(chan []byte, 100)
+	buf := <-clientInputs
+	mh, ok := p.handlers[path]
+
+	reply, stopServiceChan, err := func() (interface{}, chan bool, error) {
+		if !ok {
+			err := xerrors.New("the requested message hasn't been " +
+				"registered: " + path)
+			log.Error(err)
+			return nil, nil, err
+		}
+
+		msg := reflect.New(mh.msgType).Interface()
+		err := protobuf.DecodeWithConstructors(buf, msg,
+			network.DefaultConstructors(p.Context.server.Suite()))
+		if err != nil {
+			return nil, nil, xerrors.Errorf("decoding: %v", err)
+		}
+
+		return callInterfaceFunc(mh.handler, msg, mh.streaming)
+	}()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// This goroutine is responsible for listening on the service channel,
+	// decoding the messages and then forwarding them to the streaming tunnel,
+	// which should then forward the message to the client.
+	go func() {
```

Comment:
Instead of running the go routine right away thus forcing you to listen for the client input (which could hang, be closed, etc.. so more stuff to manage in two places), why don't you start by listening client inputs and spinning up a go routine for the service processing ? You can wait for completion if you want to be sure to be synchronous.

Reply:
If I understand you well, this is already the case: In L453 I wait for the client to send a first message, then I launch the go routines.

Reply:
Updated so that each new request from the client creates a new goroutine.